### PR TITLE
[feat]: [CDS-52508]: support for service propagation from stage template

### PR DESCRIPTION
**Pull request by tchaurasiya**
## Describe your changes
This is to support useFromStage feature for templates
* UI changes are pending user can use the feature from yaml
* A stage can only use the service configuration of other stage having same deployment type (from ui this validation would be for stages having same template
* OneOf(useFromStage,serviceRef) has been removed, instead that logic has been shifted to plan creation

More details on tech spec:
https://harness.atlassian.net/wiki/spaces/CDNG/pages/21285797928/Tech+Spec+Propagate+service+from+stage+feature+for+Templates

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43120)

### CODE REVIEW

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/service/ServiceAllInOnePlanCreatorUtils.java` line 88 at r3](https://reviewable.io/reviews/harness/harness-core/43120#-NN5iNss12DDHjhrxL9g:-NN5iNss12DDHjhrxL9h:btxzlof) ([raw file](https://github.com/harness/harness-core/blob/9f7deeaecfd9069371a718bd53b34b5aeac367ff/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/service/ServiceAllInOnePlanCreatorUtils.java#L88)):*
will this handle the below scenario

>> pipeline yaml

service:
    serviceRef: <+input>
    useFromStage: <+input>


>> runtime input yaml
 service:
    serviceRef: my_svc


Basically, after merging what is the value of `serviceYamlV2.getUseFromStage()`

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 5 of 7 files at r1, 4 of 5 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43120)* status: Some of the checks are too old. 11 file(s) to review. 0 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/service/ServiceAllInOnePlanCreatorUtils.java` line 88 at r3](https://reviewable.io/reviews/harness/harness-core/43120#-NN5iNss12DDHjhrxL9g:-NN8Ipz23BoOdBv5UIfG:b99h83d) ([raw file](https://github.com/harness/harness-core/blob/9f7deeaecfd9069371a718bd53b34b5aeac367ff/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/service/ServiceAllInOnePlanCreatorUtils.java#L88)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

will this handle the below scenario

>> pipeline yaml

service:
    serviceRef: <+input>
    useFromStage: <+input>


>> runtime input yaml
 service:
    serviceRef: my_svc


Basically, after merging what is the value of `serviceYamlV2.getUseFromStage()`
</blockquote></details>

Yes this is handled.
serviceYamlV2.getUseFromStage().getValue() comes out to be null
And serviceYamlV2.getUseFromStage().getExpressionValue() comes out to be <+input>

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43120)* status: Some of the checks are too old. 11 file(s) to review. 1 of 2 approvals obtained.

**Review by hingerabhinav — APPROVED** [review]
> Reviewed 3 of 7 files at r1, 4 of 5 files at r2, 1 of 2 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43120)* status: 3 file(s) to review.

**Comment by tchaurasiya** [comment]
trigger messagecheck

**Review by utkarshchoubey-harness — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43120)* status: 3 file(s) to review.

**Review by utkarshchoubey-harness — APPROVED** [review]
> Reviewed 3 of 7 files at r1, 4 of 5 files at r2, 1 of 2 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43120)* status: :shipit: complete! ✓
